kernel/kernel-defaults/src/main/java/io/delta/kernel/defaults/internal/types/DataTypeParser.java

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.defaults.internal.types;
+
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+import io.delta.kernel.types.*;
+import static io.delta.kernel.internal.util.Preconditions.checkArgument;
+
+/**
+ * Parses Delta data types based on the
+ * <a href="https://github.com/delta-io/delta/blob/master/PROTOCOL.md#primitive-types">
+ *     serialization rules </a> outlined in the Delta Protocol.
+ */
+public class DataTypeParser {
+
+    private DataTypeParser() {}
+
+    public static StructType parseSchema(JsonNode json) {
+        DataType parsedType = parseDataType(json);
+        if (parsedType instanceof StructType) {
+            return (StructType) parsedType;
+        } else {
+            throw new IllegalArgumentException(String.format(
+                "Could not parse the following JSON as a valid StructType:\n%s", json));
+        }
+    }
+
+    /**
+     * Parses a Delta data type from JSON. Data types can either be serialized as strings (for
+     * primitive types) or as objects (for complex types).
+     *
+     * For example:
+     * <pre>
+     * // Map type field is serialized as:
+     * {
+     *   "name" : "f",
+     *   "type" : {
+     *     "type" : "map",
+     *     "keyType" : "string",
+     *     "valueType" : "string",
+     *     "valueContainsNull" : true
+     *   },
+     *   "nullable" : true,
+     *   "metadata" : { }
+     * }
+     *
+     * // Integer type field serialized as:
+     * {
+     *   "name" : "a",
+     *   "type" : "integer",
+     *   "nullable" : false,
+     *   "metadata" : { }
+     * }
+     * </pre>
+     */
+    protected static DataType parseDataType(JsonNode json) {
```

Comment:
Why `protected static`? Who is extending this?

https://stackoverflow.com/questions/24289070/why-we-should-not-use-protected-static-in-java

Reply:
for testing, made it package private